testCheckResourceAttrSet() is broken for all but TypeString attributes

I recently spent about 12 hours trying to find a bug in my code which would result in the testing function `testCheckResourceAttrSet()` reporting a certain resource attribute was not set.

In fact, the attribute WAS set, and the function contains a bug; when you try to check an attribute which is not a simple `map[string]string` and instead, eg. `[]map[string]interface{}`, the function will fail.

The proposal here is to add a comment to the function so anyone using autocomplete in the future is aware this function should be used for string attributes only. 

To verify this is the case, I wrote a simple `panic()` test inside my resource code:
```go
if dashs, ok := d.GetOk("dashboards"); ok {
  panic(fmt.Errorf("dashboards attr = %#v", dashs))
}
```

this **guaranteed** the test would exist in a stack trace if the resource attribute `dashboards` was set; it did. Removing this small bit of code resulted in the test framework error:

```console
?       github.com/grafana/terraform-provider-grafana   [no test files]
=== RUN   TestAccDataSourceDashboardsAllAndByFolderID
    data_source_dashboards_test.go:38: Step 1/1 error: Check failed: Check 5/9 error: data.grafana_dashboards.all: Attribute 'dashboards' expected to be set
--- FAIL: TestAccDataSourceDashboardsAllAndByFolderID (3.46s)
```

--- a/helper/resource/testing.go
+++ b/helper/resource/testing.go
@@ -99,9 +99,9 @@ func AddTestSweepers(name string, s *Sweeper) {
 //
 // Sweeper flags added to the "go test" command:
 //
-// -sweep: Comma-separated list of locations/regions to run available sweepers.
-// -sweep-allow-failues: Enable to allow other sweepers to run after failures.
-// -sweep-run: Comma-separated list of resource type sweepers to run. Defaults
+//     -sweep: Comma-separated list of locations/regions to run available sweepers.
+//     -sweep-allow-failues: Enable to allow other sweepers to run after failures.
+//     -sweep-run: Comma-separated list of resource type sweepers to run. Defaults
 //             to all sweepers.
 //
 // Refer to the Env prefixed constants for environment variables that further
@@ -573,16 +573,16 @@ func ParallelTest(t testing.T, c TestCase) {
 // This function will automatically find or install Terraform CLI into a
 // temporary directory, based on the following behavior:
 //
-// - If the TF_ACC_TERRAFORM_PATH environment variable is set, that Terraform
-//   CLI binary is used if found and executable. If not found or executable,
-//   an error will be returned unless the TF_ACC_TERRAFORM_VERSION environment
-//   variable is also set.
-// - If the TF_ACC_TERRAFORM_VERSION environment variable is set, install and
-//   use that Terraform CLI version.
-// - If both the TF_ACC_TERRAFORM_PATH and TF_ACC_TERRAFORM_VERSION environment
-//   variables are unset, perform a lookup for the Terraform CLI binary based
-//   on the operating system PATH. If not found, the latest available Terraform
-//   CLI binary is installed.
+//     - If the TF_ACC_TERRAFORM_PATH environment variable is set, that
+//        Terraform CLI binary is used if found and executable. If not found or
+//        executable, an error will be returned unless the
+//        TF_ACC_TERRAFORM_VERSION environment variable is also set.
+//     - If the TF_ACC_TERRAFORM_VERSION environment variable is set, install
+//        and use that Terraform CLI version.
+//     - If both the TF_ACC_TERRAFORM_PATH and TF_ACC_TERRAFORM_VERSION
+//        environment variables are unset, perform a lookup for the Terraform
+//        CLI binary based on the operating system PATH. If not found, the
+//        latest available Terraform CLI binary is installed.
 //
 // Refer to the Env prefixed constants for additional details about these
 // environment variables, and others, that control testing functionality.
@@ -763,13 +763,13 @@ func ComposeAggregateTestCheckFunc(fs ...TestCheckFunc) TestCheckFunc {
 // Use this as a last resort when a more specific TestCheckFunc cannot be
 // implemented, such as:
 //
-// - TestCheckResourceAttr: Equality checking of non-TypeSet state value.
-// - TestCheckResourceAttrPair: Equality checking of non-TypeSet state value,
-//                              based on another state value.
-// - TestCheckTypeSet*: Equality checking of TypeSet state values.
-// - TestMatchResourceAttr: Regular expression checking of non-TypeSet state
-//                          value.
-// - TestMatchTypeSet*: Regular expression checking on TypeSet state values.
+//     - TestCheckResourceAttr: Equality checking of non-TypeSet state value.
+//     - TestCheckResourceAttrPair: Equality checking of non-TypeSet state
+//        value, based on another state value.
+//     - TestCheckTypeSet*: Equality checking of TypeSet state values.
+//     - TestMatchResourceAttr: Regular expression checking of non-TypeSet
+//        state value.
+//     - TestMatchTypeSet*: Regular expression checking on TypeSet state values.
 //
 // For managed resources, the name parameter is combination of the resource
 // type, a period (.), and the name label. The name for the below example
@@ -789,8 +789,9 @@ func ComposeAggregateTestCheckFunc(fs ...TestCheckFunc) TestCheckFunc {
 // attribute. Use the following special key syntax to inspect underlying
 // values of a list or map attribute:
 //
-// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element
-// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element
+//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
+//        value
 //
 // While it is possible to check nested attributes under list and map
 // attributes using the special key syntax, checking a list, map, or set
@@ -851,19 +852,20 @@ func testCheckResourceAttrSet(is *terraform.InstanceState, name string, key stri
 // attribute. Use the following special key syntax to inspect list, map, and
 // set attributes:
 //
-// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
-//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
-//              for sets.
-// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
-// - .#: Number of elements in list or set
-// - .%: Number of elements in map
+//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element.
+//        Use the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//        for sets.
+//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
+//        value.
+//     - .#: Number of elements in list or set.
+//     - .%: Number of elements in map.
 //
 // The value parameter is the stringified data to check at the given key. Use
 // the following attribute type rules to set the value:
 //
-// - Boolean: "false" or "true".
-// - Float/Integer: Stringified number, such as "1.2" or "123".
-// - String: No conversion necessary.
+//     - Boolean: "false" or "true".
+//     - Float/Integer: Stringified number, such as "1.2" or "123".
+//     - String: No conversion necessary.
 func TestCheckResourceAttr(name, key, value string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
 		is, err := primaryInstanceState(s, name)
@@ -940,8 +942,9 @@ func testCheckResourceAttr(is *terraform.InstanceState, name string, key string,
 // attribute. Use the following special key syntax to inspect underlying
 // values of a list or map attribute:
 //
-// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element
-// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element.
+//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
+//        value.
 //
 // While it is possible to check nested attributes under list and map
 // attributes using the special key syntax, checking a list, map, or set
@@ -1015,12 +1018,13 @@ func testCheckNoResourceAttr(is *terraform.InstanceState, name string, key strin
 // attribute. Use the following special key syntax to inspect list, map, and
 // set attributes:
 //
-// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
-//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
-//              for sets.
-// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
-// - .#: Number of elements in list or set
-// - .%: Number of elements in map
+//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element.
+//        Use the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//        for sets.
+//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
+//        value.
+//     - .#: Number of elements in list or set.
+//     - .%: Number of elements in map.
 //
 // The value parameter is a compiled regular expression. A typical pattern is
 // using the regexp.MustCompile() function, which will automatically ensure the
@@ -1110,12 +1114,13 @@ func TestCheckModuleResourceAttrPtr(mp []string, name string, key string, value 
 // attribute. Use the following special key syntax to inspect list, map, and
 // set attributes:
 //
-// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
-//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
-//              for sets.
-// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
-// - .#: Number of elements in list or set
-// - .%: Number of elements in map
+//     - .{NUMBER}: List value at index, e.g. .0 to inspect the first element.
+//        Use the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//        for sets.
+//     - .{KEY}: Map value at key, e.g. .example to inspect the example key
+//        value.
+//     - .#: Number of elements in list or set.
+//     - .%: Number of elements in map.
 func TestCheckResourceAttrPair(nameFirst, keyFirst, nameSecond, keySecond string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSetPair(keyFirst, keySecond, func(s *terraform.State) error {
 		isFirst, err := primaryInstanceState(s, nameFirst)

--- a/helper/resource/testing.go
+++ b/helper/resource/testing.go
@@ -718,6 +718,9 @@ func testResource(c TestStep, state *terraform.State) (*terraform.ResourceState,
 //
 // As a user testing their provider, this lets you decompose your checks
 // into smaller pieces more easily.
+//
+// ComposeTestCheckFunc returns immediately on the first TestCheckFunc error.
+// To aggregrate all errors, use ComposeAggregateTestCheckFunc instead.
 func ComposeTestCheckFunc(fs ...TestCheckFunc) TestCheckFunc {
 	return func(s *terraform.State) error {
 		for i, f := range fs {
@@ -752,10 +755,47 @@ func ComposeAggregateTestCheckFunc(fs ...TestCheckFunc) TestCheckFunc {
 	}
 }
 
-// TestCheckResourceAttrSet is a TestCheckFunc which ensures a value
-// exists in state for the given name/key combination. It is useful when
-// testing that computed values were set, when it is not possible to
-// know ahead of time what the values will be.
+// TestCheckResourceAttrSet ensures any value exists in the state for the
+// given name and key combination. The opposite of this TestCheckFunc is
+// TestCheckNoResourceAttr. State value checking is only recommended for
+// testing Computed attributes and attribute defaults.
+//
+// Use this as a last resort when a more specific TestCheckFunc cannot be
+// implemented, such as:
+//
+// - TestCheckResourceAttr: Equality checking of non-TypeSet state value.
+// - TestCheckResourceAttrPair: Equality checking of non-TypeSet state value,
+//                              based on another state value.
+// - TestCheckTypeSet*: Equality checking of TypeSet state values.
+// - TestMatchResourceAttr: Regular expression checking of non-TypeSet state
+//                          value.
+// - TestMatchTypeSet*: Regular expression checking on TypeSet state values.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect underlying
+// values of a list or map attribute:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+//
+// While it is possible to check nested attributes under list and map
+// attributes using the special key syntax, checking a list, map, or set
+// attribute directly is not supported. Use TestCheckResourceAttr with
+// the special .# or .% key syntax for those situations instead.
 func TestCheckResourceAttrSet(name, key string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
 		is, err := primaryInstanceState(s, name)
@@ -789,8 +829,41 @@ func testCheckResourceAttrSet(is *terraform.InstanceState, name string, key stri
 	return nil
 }
 
-// TestCheckResourceAttr is a TestCheckFunc which validates
-// the value in state for the given name/key combination.
+// TestCheckResourceAttr ensures a specific value is stored in state for the
+// given name and key combination. State value checking is only recommended for
+// testing Computed attributes and attribute defaults.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect list, map, and
+// set attributes:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
+//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//              for sets.
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+// - .#: Number of elements in list or set
+// - .%: Number of elements in map
+//
+// The value parameter is the stringified data to check at the given key. Use
+// the following attribute type rules to set the value:
+//
+// - Boolean: "false" or "true".
+// - Float/Integer: Stringified number, such as "1.2" or "123".
+// - String: No conversion necessary.
 func TestCheckResourceAttr(name, key, value string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
 		is, err := primaryInstanceState(s, name)
@@ -844,8 +917,36 @@ func testCheckResourceAttr(is *terraform.InstanceState, name string, key string,
 	return nil
 }
 
-// TestCheckNoResourceAttr is a TestCheckFunc which ensures that
-// NO value exists in state for the given name/key combination.
+// TestCheckNoResourceAttr ensures no value exists in the state for the
+// given name and key combination. The opposite of this TestCheckFunc is
+// TestCheckResourceAttrSet. State value checking is only recommended for
+// testing Computed attributes and attribute defaults.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect underlying
+// values of a list or map attribute:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+//
+// While it is possible to check nested attributes under list and map
+// attributes using the special key syntax, checking a list, map, or set
+// attribute directly is not supported. Use TestCheckResourceAttr with
+// the special .# or .% key syntax for those situations instead.
 func TestCheckNoResourceAttr(name, key string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
 		is, err := primaryInstanceState(s, name)
@@ -892,8 +993,39 @@ func testCheckNoResourceAttr(is *terraform.InstanceState, name string, key strin
 	return nil
 }
 
-// TestMatchResourceAttr is a TestCheckFunc which checks that the value
-// in state for the given name/key combination matches the given regex.
+// TestMatchResourceAttr ensures a value matching a regular expression is
+// stored in state for the given name and key combination. State value checking
+// is only recommended for testing Computed attributes and attribute defaults.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect list, map, and
+// set attributes:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
+//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//              for sets.
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+// - .#: Number of elements in list or set
+// - .%: Number of elements in map
+//
+// The value parameter is a compiled regular expression. A typical pattern is
+// using the regexp.MustCompile() function, which will automatically ensure the
+// regular expression is supported by the Go regular expression handlers during
+// compilation.
 func TestMatchResourceAttr(name, key string, r *regexp.Regexp) TestCheckFunc {
 	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
 		is, err := primaryInstanceState(s, name)
@@ -935,6 +1067,9 @@ func testMatchResourceAttr(is *terraform.InstanceState, name string, key string,
 // TestCheckResourceAttrPtr is like TestCheckResourceAttr except the
 // value is a pointer so that it can be updated while the test is running.
 // It will only be dereferenced at the point this step is run.
+//
+// Refer to the TestCheckResourceAttr documentation for more information about
+// setting the name, key, and value parameters.
 func TestCheckResourceAttrPtr(name string, key string, value *string) TestCheckFunc {
 	return func(s *terraform.State) error {
 		return TestCheckResourceAttr(name, key, *value)(s)
@@ -949,8 +1084,38 @@ func TestCheckModuleResourceAttrPtr(mp []string, name string, key string, value 
 	}
 }
 
-// TestCheckResourceAttrPair is a TestCheckFunc which validates that the values
-// in state for a pair of name/key combinations are equal.
+// TestCheckResourceAttrPair ensures value equality in state between the first
+// given name and key combination and the second name and key combination.
+// State value checking is only recommended for testing Computed attributes
+// and attribute defaults.
+//
+// For managed resources, the name parameter is combination of the resource
+// type, a period (.), and the name label. The name for the below example
+// configuration would be "myprovider_thing.example".
+//
+//     resource "myprovider_thing" "example" { ... }
+//
+// For data sources, the name parameter is a combination of the keyword "data",
+// a period (.), the data source type, a period (.), and the name label. The
+// name for the below example configuration would be
+// "data.myprovider_thing.example".
+//
+//     data "myprovider_thing" "example" { ... }
+//
+// The first and second names may use any combination of managed resources
+// and/or data sources.
+//
+// The key parameter is an attribute path in Terraform CLI 0.11 and earlier
+// "flatmap" syntax. Keys start with the attribute name of a top-level
+// attribute. Use the following special key syntax to inspect list, map, and
+// set attributes:
+//
+// - .{NUMBER}: List value at index, e.g. .0 to inspect the first element. Use
+//              the TestCheckTypeSet* and TestMatchTypeSet* functions instead
+//              for sets.
+// - .{KEY}: Map value at key, e.g. .example to inspect the example key value
+// - .#: Number of elements in list or set
+// - .%: Number of elements in map
 func TestCheckResourceAttrPair(nameFirst, keyFirst, nameSecond, keySecond string) TestCheckFunc {
 	return checkIfIndexesIntoTypeSetPair(keyFirst, keySecond, func(s *terraform.State) error {
 		isFirst, err := primaryInstanceState(s, nameFirst)

--- a/helper/resource/testing_example_test.go
+++ b/helper/resource/testing_example_test.go
@@ -1,0 +1,282 @@
+package resource_test
+
+import (
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func ExampleComposeAggregateTestCheckFunc() {
+	// This function is typically implemented in a TestStep type Check field.
+	// Any TestCheckFunc and number of TestCheckFunc may be used within the
+	// function parameters. Any errors are combined and displayed together.
+	resource.ComposeAggregateTestCheckFunc(
+		resource.TestCheckResourceAttr("example_thing.test", "example_attribute1", "one"),
+		resource.TestCheckResourceAttr("example_thing.test", "example_attribute2", "two"),
+	)
+}
+
+func ExampleTestCheckNoResourceAttr() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_string_attribute = "test-value"
+	//     }
+	//
+	// The following TestCheckNoResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckNoResourceAttr("example_thing.test", "non_existent_attribute")
+}
+
+func ExampleTestCheckResourceAttr_typeBool() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_bool_attribute = true
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttr("example_thing.test", "example_bool_attribute", "true")
+}
+
+func ExampleTestCheckResourceAttr_typeFloat() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_float_attribute = 1.2
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttr("example_thing.test", "example_float_attribute", "1.2")
+}
+
+func ExampleTestCheckResourceAttr_typeInt() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_int_attribute = 123
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttr("example_thing.test", "example_int_attribute", "123")
+}
+
+func ExampleTestCheckResourceAttr_typeListAttribute() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_list_attribute = ["value1", "value2", "value3"]
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+
+	// Verify the list attribute contains 3 and only 3 elements
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_attribute.#", "3")
+
+	// Verify each list attribute element value
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_attribute.0", "value1")
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_attribute.1", "value2")
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_attribute.2", "value3")
+}
+
+func ExampleTestCheckResourceAttr_typeListBlock() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_list_block {
+	//         example_string_attribute = "test-nested-value"
+	//       }
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+
+	// Verify the list block contains 1 and only 1 definition
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_block.#", "1")
+
+	// Verify a first list block attribute value
+	resource.TestCheckResourceAttr("example_thing.test", "example_list_block.0.example_string_attribute", "test-nested-value")
+}
+
+func ExampleTestCheckResourceAttr_typeMap() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_map_attribute = {
+	//         key1 = "value1"
+	//         key2 = "value2"
+	//         key3 = "value3"
+	//       }
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+
+	// Verify the map attribute contains 3 and only 3 elements
+	resource.TestCheckResourceAttr("example_thing.test", "example_map_attribute.%", "3")
+
+	// Verify each map attribute element value
+	resource.TestCheckResourceAttr("example_thing.test", "example_map_attribute.key1", "value1")
+	resource.TestCheckResourceAttr("example_thing.test", "example_map_attribute.key2", "value2")
+	resource.TestCheckResourceAttr("example_thing.test", "example_map_attribute.key3", "value3")
+}
+
+func ExampleTestCheckResourceAttr_typeString() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_string_attribute = "test-value"
+	//     }
+	//
+	// The following TestCheckResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttr("example_thing.test", "example_string_attribute", "test-value")
+}
+
+func ExampleTestCheckResourceAttrPair() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test1" {
+	//       example_string_attribute = "test-value"
+	//     }
+	//
+	//     resource "example_thing" "test2" {
+	//       example_string_attribute = example_thing.test1.example_string_attribute
+	//     }
+	//
+	// The following TestCheckResourceAttrPair can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttrPair(
+		"example_thing.test1",
+		"example_string_attribute",
+		"example_thing.test2",
+		"example_string_attribute",
+	)
+}
+
+func ExampleTestCheckResourceAttrSet() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_string_attribute = "test-value"
+	//     }
+	//
+	// The following TestCheckResourceAttrSet can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestCheckResourceAttrSet("example_thing.test", "example_string_attribute")
+}
+
+func ExampleTestMatchResourceAttr() {
+	// This function is typically implemented in a TestStep type Check field,
+	// wrapped with ComposeAggregateTestCheckFunc to combine results from
+	// multiple checks.
+	//
+	// Given the following example configuration:
+	//
+	//     resource "example_thing" "test" {
+	//       example_string_attribute = "test-value"
+	//     }
+	//
+	// The following TestMatchResourceAttr can be written to assert against
+	// the expected state values.
+	//
+	// NOTE: State value checking is only necessary for Computed attributes,
+	//       as the testing framework will automatically return test failures
+	//       for configured attributes that mismatch the saved state, however
+	//       this configuration and test is shown for illustrative purposes.
+	resource.TestMatchResourceAttr("example_thing.test", "example_string_attribute", regexp.MustCompile(`^test-`))
+}

